File: Week03/Code/Vectorize2.R
# Runs the stochastic (with gaussian fluctuations) Ricker Eqn .

rm(list=ls())

stochrick<-function(p0=runif(1000,.5,1.5),r=1.2,K=1,sigma=0.2,numyears=100)
{
  #initialize
  N<-matrix(NA,numyears,length(p0))
  N[1,]<-p0
  
  for (pop in 1:length(p0)) #loop through the populations
  {
    for (yr in 2:numyears) #for each pop, loop through the years
    {
      N[yr,pop]<-N[yr-1,pop]*exp(r*(1-N[yr-1,pop]/K)+rnorm(1,0,sigma))
    }
  }
 return(N)

}

## My function
stochrick.m<-function(p0=runif(1000,.5,1.5),r=1.2,K=1,sigma=0.2,numyears=100)
{
  #initialize
  N<-as.data.frame(matrix(NA,numyears,length(p0))) ## content, row, col
  N[1,]<-p0
  for(i in 2:numyears){
    N[i,]<-apply(N[i-1,],2,sum)
  }
  # for (pop in 1:length(p0)) #loop through the populations
  # {
  #   for (yr in 2:numyears) #for each pop, loop through the years
  #   {
  #     N[yr,pop]<-N[yr-1,pop]*exp(r*(1-N[yr-1,pop]/K)+rnorm(1,0,sigma))
  #   }
  # }
  return(N)
  
}

# Now write another function called stochrickvect that vectorizes the above to the extent possible, with improved performance:

## comparison
print("Vectorized Stochastic Ricker takes:")
print(system.time(res2<-stochrick()))
print("My Vectorized Stochastic Ricker takes:")
print(system.time(res3<-stochrick.m()))

File: Week3/Code/Vectorize2.py
#!/bin/env python3

## imports
import random
import time
import numpy
import math

## variables

##func
def stochrick(p0=list([random.random()+.5 for i in range(1000)]),r=1.2,K=1,sigma=.2,numyears=100):
    ## initialize
    N=numpy.zeros((numyears,len(p0)))
    N[0,]=p0
    for pop in range(len(p0)):
        for yr in range(1,numyears):
            N[yr][pop]=N[yr-1][pop]*math.exp(r*(1-N[yr-1][pop]/K)+float(numpy.random.normal(0,sigma,1)))
    return N

def stochrick_m(p0=1000,r=1.2,K=1,sigma=0.2,numyears=100):
    ## initialize
    N=numpy.zeros((numyears,p0))
    N[0]=list([random.random()+.5 for i in range(p0)])
    ## calculate
    for pop in range(p0):
        for yr in range(1,numyears):
            N[yr][pop]=N[yr-1][pop]*math.exp(r*(1-N[yr-1][pop]/K)+float(numpy.random.normal(0,sigma,1)))
    return N

print("My Vectorized Stochastic Ricker takes:")
start=time.time()
stochrick_m()
print("%s" %(round(time.time()-start,3)))

print("Vectorized Stochastic Ricker takes:")
start=time.time()
stochrick()
print("%s" %(round(time.time()-start,3)))